lineapy/graph_reader/artifact_collection.py

```diff
@@ -46,121 +39,141 @@ class ArtifactCollection:
-            # Retrieve artifact
-            art = LineaArtifact.get_artifact_from_def(self.db, art_def)
-            self.art_name_to_node_id[art_def["artifact_name"]] = art._node_id
-            self.node_id_to_session_id[art._node_id] = art._session_id
-            # Put artifact in the right session group
-            artifacts_by_session[art._session_id].append(art)
-            # Record session_id of an artifact
-            self.artifact_session[art.name] = art._session_id
-            self.artifact_session[slugify(art.name)] = art._session_id
 
-        pre_calculated_artifacts: Dict[str, LineaArtifact] = {}
-        for art_entry in reuse_pre_computed_artifacts:
-            art_def = get_lineaartifactdef(art_entry=art_entry)
-            # Check no two reuse pre-computed artifacts could have the same name
-            # Otherwise will confuse which one to use
-            if art_def["artifact_name"] in pre_calculated_artifacts.keys():
-                msg = (
-                    "Duplicated reuse_pre_computed_artifacts names detected in "
-                    + f"{list(pre_calculated_artifacts.keys())}"
-                )
-                logger.error(msg)
-                raise KeyError(msg)
-            # Retrieve artifact
-            art = LineaArtifact.get_artifact_from_def(self.db, art_def)
-            pre_calculated_artifacts[art.name] = art
+        reuse_pre_computed_artifact_defs = [
+            get_lineaartifactdef(art_entry=art_entry)
+            for art_entry in reuse_pre_computed_artifacts
+        ]
+        # Retrieve reuse precomputed artifact objects and group them by session ID
+        self.pre_computed_artifacts_by_session = (
+            self._get_artifacts_grouped_by_session(
+                reuse_pre_computed_artifact_defs
+            )
+        )
 
         # For each session, construct SessionArtifacts object
-        for session_id, session_artifacts in artifacts_by_session.items():
+        self.session_artifacts: Dict[LineaID, SessionArtifacts] = {}
+        for (
+            session_id,
+            session_artifacts,
+        ) in self.target_artifacts_by_session.items():
+            # TODO: LIN-653, LIN-640 only collect matched pre_computed artifacts in this Session
+            # instead of grabbing all Artifacts as currently required by SessionArtifact
+            pre_calculated_artifacts: List[LineaArtifact] = sum(
+                self.pre_computed_artifacts_by_session.values(), []
+            )
+
             self.session_artifacts[session_id] = SessionArtifacts(
                 self.db,
                 session_artifacts,
                 input_parameters=input_parameters,
                 reuse_pre_computed_artifacts=pre_calculated_artifacts,
             )
 
+        # TODO: LIN-653, LIN-640
+        # Add some type of validation that all re-use compute artifacts can be matched to
+        # a target artifact before the creation of SessionArtifacts.
```

Comment:
Curious about how you can find out matched pre_computed artifacts before you create SessionArtifact. The graph in the SessionArtifact is not the entire session graph.

Reply:
Not really sure. I agree SessionArtifact has some chicken and the egg problem where you need to "create" the slice that defines the SessionArtifact before you can tell if  pre_computed artifacts can be in the SessionArtifact. 

We can move this logic out of SessionArtifacts, but then the constructor for SessionArtifact would be empty and the class would be useless ...

I myself need a bit more time to understand how the constructor for SessionArtifacts works before revisiting this.